src/index.d.ts

```diff
@@ -0,0 +1,158 @@
+    setValues: Dispatch<SetStateAction<T>>
+    runValidation: () => void
+    getFieldValue: <V = any>(fieldPath: keyof T | string) => V
+    getFieldMeta: (
+      fieldPath: keyof T | string
+    ) => { error: string | null; [k: string]: any }
+    setFieldValue: <V = any>(
+      fieldPath: keyof T | string,
+      updater: SetValue<V>,
+      options?: { isTouched: boolean }
+    ) => void
+    setFieldMeta: (fieldPath: keyof T | string, value: SetFieldMeta) => void
+    pushFieldValue: <V = any>(fieldPath: keyof T | string, value: V) => void
+    insertFieldValue: <V = any>(
+      fieldPath: keyof T | string,
+      insertIndex: number,
+      value: V
+    ) => void
+    removeFieldValue: (
+      fieldPath: keyof T | string,
+      removalIndex: number
+    ) => void
+    swapFieldValues: (
+      fieldPath: keyof T | string,
+      firstIndex: number,
+      secondIndex: number
+    ) => void
+  }
+
+  interface FormOptions<T> {
+    defaultValues: T
+    onSubmit: (values: T, instance: FormInstance<T>) => OptionalPromise<void>
+    validate: (
+      values: Partial<T>,
+      instance: FormInstance<T>
+    ) => OptionalPromise<ValidatorReturn>
+    validatePristine?: boolean
+    debugForm: boolean
+  }
+
+  export function useForm<T extends {} = {}>(
+    options?: Partial<FormOptions<T>>
+  ): FormInstance<T>
+
+  export function useFormContext<T extends {} = {}>(): FormInstance<T>
+
+  interface FieldMeta {
+    error: string | false
+    isTouched: boolean
+    [k: string]: any
+  }
+
+  type SetValue<S> = S | ((prevState: S) => S)
+  type SetFieldMeta =
+    | Partial<FieldMeta>
+    | ((previousMeta: FieldMeta) => FieldMeta)
+
+  interface FieldInstance<T = any, F = any> {
+    form: FormInstance<F>
+    fieldName: string
+    value: T
+    meta: FieldMeta
+    FieldScope?: ComponentType<Provider<any>>
+    debounce: Debounce
+    runValidation: () => void
+    getInputProps: (
+      value: Partial<HTMLProps<HTMLInputElement>> &
+        Pick<HTMLProps<HTMLInputElement>, 'onSubmit'> &
+        Pick<HTMLProps<HTMLInputElement>, 'onBlur'>
+    ) => {
+      value: T
+      onChange: ChangeEventHandler<HTMLInputElement>
+      onBlur: FocusEventHandler<HTMLInputElement>
+    } & Partial<HTMLProps<HTMLInputElement>>
+    setValue: <V = any>(
+      updater: SetValue<V>,
+      options?: { isTouched: boolean }
+    ) => void
+    setMeta: (value: SetFieldMeta) => void
+    pushValue: <V = any>(value: V) => void
+    insertValue: <V = any>(insertIndex: number, value: V) => void
+    removeValue: (removalIndex: number) => void
+    swapValues: (firstIndex: number, secondIndex: number) => void
+  }
+
+  interface FieldOptions<T> {
+    defaultValue?: T
+    defaultError?: string
+    defaultIsTouched?: boolean
+    defaultMeta?: FieldMeta
+    validate?: (
+      value: T,
+      instance: FieldInstance
+    ) => OptionalPromise<ValidatorReturn>
+    filterValue?: <T = any>(values: T, instance: FieldInstance) => T
```

Comment:
- Why is there another `T` here? T comes in via `FieldOptions`. If its the same, remove the `<T = any>` here, otherwise rename.
- Type is not passed to `FieldInstance<T = any, F = any>`. This probably can be forwarded? `FieldInstance<T>`.

Reply:
Good catch, thanks!